log4j-slf4j-impl/src/main/java/org/apache/logging/slf4j/Log4jMarkerFactory.java

```diff
@@ -113,26 +125,18 @@ public boolean exists(final String name) {
         return markerMap.containsKey(name);
     }
 
-    /**
-     * Log4j does not support detached Markers. This method always returns false.
-     * @param name The Marker name.
-     * @return {@code false}
-     */
     @Override
     public boolean detachMarker(final String name) {
-        return false;
+        if (name == null) {
+            return false;
+        }
+        detachedMarkers.add(name);
+        return true;
```

Comment:
I thought the point of detached markers was so that we can create arbitrarily many without worrying about the impact on heap -- if we hold a collection `detachedMarkers` strings, that is no longer true, and likely results in a memory leak.
I would suggest against supporting the `boolean detachMarker(final String name) {` method, at least until we have a well understood use-case for it.

Reply:
I do have a use case :-) 

I will call this method *once* with "STRUCTURE_ARG". This is needed to avoid caching of the incoming `Marker` instance by name (all my markers are value-based classes). 

In `Log4jLogger` there is the following code:

```java
@Override
public void info(final Marker marker, final String s) {
    logger.logIfEnabled(FQCN, Level.INFO, getMarker(marker), s);
}
```

the ultimate goal of `detachedMarkers` is to skip the last line of the `getMarker` method :

```
    public Marker getMarker(final Marker marker) {
        if (detachedMarkers.contains(name)) {
            return marker;
        }
        final Marker m = markerMap.get(name);
        if (m != null) {
            return m;
        }
        return addMarkerIfAbsent(name, convertMarker(marker));  <-- I want to skip this processing
```

PS
Of course, it could be abused by calling it with millions of unique strings but that is true also for the non-detached version...

Reply:
I think the `Log4jLogger` marker conversion fallback (when the slf4j marker implementation is not a Log4jMarker) should use a simple wrapper around the input instance without any caching. Logging would incur object allocation, but this is a special case where markers don't come from the marker factory.
https://github.com/apache/logging-log4j2/blob/f72100df0decc9bda96b4d769822c4e48b2848fc/log4j-slf4j-impl/src/main/java/org/apache/logging/slf4j/Log4jLogger.java#L379-L388

What do you think?

Reply:
@carterkozak yes, I'm already doing this with custom `org.slf4j.impl.StaticMarkerBinder` that is wrapping the incoming marker with `Log4jMarker` (to go via line 383 instead of line 386). 

Problem is that now our applications are sensitive to the order of the jars in the classpath: my binding must come *before* log4j-slf4j-impl. Do you see another way perhaps? (and thank you for the nice review by the way!)